SSP Surface Hydrostatic Force Derivation

bmaclach opened the pull request:
This PR adds the derivation for the surface hydrostatic force DD in SSP according to #1277.

Some notes for @smiths: I added a DD for the concept of weight as suggested, though I didn't go further back in the derivation than that (i.e. I didn't show how weight is derived from Newton's second law). As you thought, Newton's second law is indeed already defined in GamePhysics, but it is a Theoretical Model, and DD's cannot reference Theoretical Models according to the graph you made (https://github.com/JacquesCarette/Drasil/issues/1029#issuecomment-434077273).

bmaclach commented:
@smiths I left the Source field for the new DD for the concept of weight blank. What sources do we prefer to use for very general, common definitions like that one? I noticed in SWHS that the source for the definition of latent heat was the Wikipedia page for latent heat. Would it be all right to do a similar thing here, or do we want to minimize the use of Wikipedia as a source?

smiths commented on code/stable/ssp/SRS/SSP_SRS.tex:
```diff
@@ -144,8 +144,14 @@ \subsection{Table of Symbols}
 \\
+${\mathbf{V}_{wat}}$ & Volumes of Water: amount of space occupied by water for each slice & $\text{m}^{3}$
+\\
+$W$ & Weight: The gravitational force acting on an object & N
```
I thought for a second that having unbold W and bold W might be confusing, but the more I thought about it the more I'm okay with it.  unbold W on its own is not a component of bold W.  The components of bold W need an index to represent them, so there is no confusion between unbold W and bold W.

JacquesCarette commented:
Changes are good, but a merge still needs to happen.

smiths commented:
As far as the source field goes, I'm fine with reference to Wikipedia.  Technically, it is supposed to be better to cite a physical source, but I think we should relax this for common physics concepts.  We could cite a physics book I have on my shelf, or that is in the library, and nobody is ever going to follow up on the source.  If we cite Wikipedia, future users might actually go to the source to learn something.  This benefit outweighs, in my mind, the danger that Wikipedia may have false information.  When it comes to physical concepts, I've been impressed with the quality of the Wikipedia entries.

bmaclach commented:
@smiths We can discuss at tomorrow's meeting, but just to record my responses to your comments:

- The problem where weight is not indexed on the left-hand side of the equation is a known issue with Data Definitions where we can't modify the left-hand side (#359)
- I pushed a commit fixing the problem with γdry being referenced in the wrong DD, thanks for catching that.
- For your suggestion that "U_t = V_wat * gamma_w should be written for one slice, not for all slices", would it be enough to index the vector symbols in the equation (so they would still be bold but with an index)? Or do you want them unbolded and unindexed? Indexing them is easy, unbolding them not so. It would require defining 2 new unitals for the non-vector versions of these quantities.
- I agree with adding the new assumption.
- I agree that I'd like to consolidate the DDs if we can. The drawback is that it might look very messy with all of the different cases in a single equation.

JacquesCarette commented:
Re: Wikipedia or not - that seems like a false choice to me. I'm fine with using Wikipedia *and* many other sources.

smiths commented:
@JacquesCarette, great point about multiple sources, not just Wikipedia.

smiths commented:
Thank you for your feedback to my comments @bmaclach.  My response follows:

- The problem where weight is not indexed on the left-hand side of the equation is a known issue with Data Definitions where we can't modify the left-hand side (#359)

I forgot about this issue.  Thank you for the reminder.  I'm glad it is officially on our "to do" list.

- I pushed a commit fixing the problem with γdry being referenced in the wrong DD, thanks for catching that.

Great!

- For your suggestion that "U_t = V_wat * gamma_w should be written for one slice, not for all slices", would it be enough to index the vector symbols in the equation (so they would still be bold but with an index)? Or do you want them unbolded and unindexed? Indexing them is easy, unbolding them not so. It would require defining 2 new unitals for the non-vector versions of these quantities.

Indexes are fine.  The symbol can remain bold.

- I agree with adding the new assumption.

Great!

- I agree that I'd like to consolidate the DDs if we can. The drawback is that it might look very messy with all of the different cases in a single equation.

With the new assumption, I think we only have two cases: 1) one of the sides has the water table above the slope, or 2) neither side has the water above the slope.  We can discuss further tomorrow (Thursday).

bmaclach commented:
I addressed all the feedback with the last batch of commits, including changing the DDs to GDs, which we decided at the meeting. Since we wanted to reference the Newton's 2nd law TM from game physics, this is the first time that we needed a TM in drasil-data to reuse between examples, so I created a new Theories/ directory in drasil-data to store it. I also put the Weight GD in there since it is very generic. If the name of the directory should be changed, or if you would prefer to store TMs and GDs in separate files, let me know.

In making the requested changes, this PR got rather large. I can try to split it up into smaller PRs if you want, but they will all be dependent and need to be merged in a certain order @JacquesCarette

JacquesCarette commented:
This passes Travis, but there's one more conflict to resolve. You may as well hold off for ~15 minutes, as I'm about to merge something from @danscime that's likely to cause more conflicts.

JacquesCarette commented:
Hey, not too bad, just 2 files have conflicts.

JacquesCarette commented:
I guess I should wait for @smiths to re-review before merging it in, as he's the one who requested changes.

smiths commented:
Yes, please wait.  I'll get to my review this afternoon.

bmaclach commented:
For point 2., I do think that due to our new assumption the condition I have is correct. If one side is above zero, we are assuming that the second side will be too. But your suggestion is less confusing and is easy to implement so that seems like a good idea. The requested changes are on their way!

smiths commented:
@bmaclach, I believe I see what you are getting at with my point 2.  Since your test includes the case where the water table is exactly at the soil surface (ywt,i = yslope,i), you are correct that the case where the i-1 side is above zero is covered, because of our assumption.  I agree with you though that my suggestion is clearer.  Since it isn't hard to change, let's go with my suggestion.  Thank you!

bmaclach commented:
@smiths Both changes have been made in the last batch of commits.

JacquesCarette commented:
master needs re-merged, and @smiths should re-review.

JacquesCarette commented:
If @smiths agrees, this looks ready to merge.

smiths commented:
The revisions look good to me, although there appears to be conflicts now.
